cpp/src/copying/contiguous_split.cu

```diff
@@ -762,6 +792,150 @@ struct size_of_helper {
   }
 };
 
+/**
+ * @brief Functor for returning the number of chunks an input buffer is being
+ * subdivided into during the repartitioning step.
+ *
+ * Note: columns types which themselves inherently have no data (strings, lists,
+ * structs) return 0.
+ */
+struct num_chunks_func {
+  thrust::pair<size_t, size_t> const* chunks;
+  __device__ size_t operator()(size_type i) const { return thrust::get<0>(chunks[i]); }
+};
+
+void copy_data(size_t total_bytes,
+               int num_bufs,
+               int num_src_bufs,
+               uint8_t const** d_src_bufs,
+               uint8_t** d_dst_bufs,
+               dst_buf_info* _d_dst_buf_info,
+               rmm::cuda_stream_view stream)
+{
+  // ideally we'd like to give each SM a similar amount of work to do so that a.) we keep all
+  // of them saturated and b.) we don't have any long-running outliers.
+  // our incoming dst_buf_info data is the exact description of what the output should look like.
+  // let's do some examination of what's being copied and potentially break things up into
+  // more pieces to parallelize better.
+  int device;
+  cudaGetDevice(&device);
+  cudaDeviceProp prop;
+  cudaGetDeviceProperties(&prop, device);
+
+  // distribute the # of chunks to be copied roughly evenly among the SMs we have
+  rmm::device_uvector<thrust::pair<size_t, size_t>> chunks(num_bufs, stream);
+
+  auto const num_sms = prop.multiProcessorCount;
+  thrust::transform(
+    rmm::exec_policy(stream),
+    _d_dst_buf_info,
+    _d_dst_buf_info + num_bufs,
+    chunks.begin(),
+    [total_bytes_f = static_cast<float>(total_bytes), num_sms] __device__(dst_buf_info const& buf) {
+      // how many chunks do we want to subdivide this buffer into
+      size_t const bytes = buf.num_elements * buf.element_size;
+      // can happen for things like lists and strings (the root columns store no data)
+      if (bytes == 0) { return thrust::pair<size_t, size_t>{1, 0}; }
+      float const fraction          = static_cast<float>(bytes) / total_bytes_f;
+      size_t const ideal_num_chunks = max(size_t{1}, static_cast<size_t>(fraction * num_sms));
+
+      // make sure chunks are padded/aligned to 64 bytes.
+      size_t const chunk_size =
+        max(size_t{split_align}, _round_up_safe(bytes / ideal_num_chunks, split_align));
+      size_t const num_chunks = _round_up_safe(bytes, chunk_size) / chunk_size;
+      return thrust::pair<size_t, size_t>{num_chunks, chunk_size};
```

Comment:
With CTAD this should work without specifying the template parameters, right? If for some reason that doesn't work with `thrust::pair` I'd recommend using `make_pair` instead.